src/Tags.elm

```diff
@@ -0,0 +1,175 @@
+module Tags exposing (State, Msg, init, update, view)
+
+import Dict
+import Html exposing (Html)
+import Html.Attributes
+import Html.Events
+import Form exposing (Form)
+import Form.Field as Field
+import Form.Input as Input
+import Form.Validate as Validate exposing (..)
+import Form.Error as Error exposing (..)
+import UserGroup exposing (Tags)
+import Errors exposing (errorString)
+
+
+type alias State =
+    { form : Form () Tags }
+
+
+type Msg
+    = FormMsg Form.Msg
+
+
+
+---- INIT ----
+
+
+init : Tags -> State
+init tags =
+    let
+        fields =
+            [ ( "tags", Field.list (List.map (Field.group << tagFields) (Dict.toList tags)) ) ]
+
+        tagFields ( name, value ) =
+            [ ( "name", Field.string name )
+            , ( "value", value |> Maybe.withDefault "" |> Field.string )
+            ]
+    in
+        { form = Form.initial fields validate }
+
+
+
+---- VALIDATION ----
+
+
+validate : Validation () Tags
+validate =
+    field "tags" (list tag)
+        |> andThen (\t -> succeed (Dict.fromList t))
+
+
+tag : Validation () (String, Maybe String)
+tag =
+    map2 Tuple.pair
+        (field "name" (string |> andThen (maxLength 32)))
+        (field "value" (oneOf [ string |> map Just
+                              , emptyString |> andThen (\_ -> succeed Nothing)
+                              ]))
+
+
+
+---- UPDATE ----
+
+
+update : Msg -> State -> ( State, Cmd Msg )
+update msg state =
+    case msg of
+        FormMsg formMsg ->
+            let
+                form = Form.update validate formMsg state.form
+            in
+                ( { state | form = form }
```

Comment:
Q: I would probably inline this to the record directly if it is the only part in `let..in`. What are the benefits of extracting this away here for you?

Reply:
Yeah, that's true. I was just trying to avoid making the line too long and hard to read.